File: project-{topic}-enhanced/src/middleware/errorHandler.middleware.ts
```typescript
import { Request, Response, NextFunction } from 'express';
import { AppError } from '../utils/AppError';
import { logger } from '../utils/logger';
import { environment } from '../config/environment';

/**
 * @file Global error handling middleware.
 *
 * This middleware catches all errors thrown by other middleware or route handlers,
 * logs them, and sends a standardized error response to the client.
 * It differentiates between operational errors (AppError) and programming errors.
 */

/**
 * Global error handling middleware for Express.
 *
 * @param {Error} err - The error object caught.
 * @param {Request} req - The Express request object.
 * @param {Response} res - The Express response object.
 * @param {NextFunction} next - The next middleware function (not used here, but required by Express).
 */
export const errorHandler = (
  err: Error,
  req: Request,
  res: Response,
  next: NextFunction
) => {
  if (res.headersSent) {
    return next(err); // Delegate to default Express error handler if headers already sent
  }

  let error = { ...err }; // Create a copy of the error object
  error.name = err.name;
  error.message = err.message;

  // Log the error
  if (err instanceof AppError) {
    logger.warn(`AppError (${err.statusCode}): ${err.message}`, {
      path: req.path,
      method: req.method,
      stack: environment.nodeEnv === 'development' ? err.stack : undefined
    });
  } else {
    logger.error(`Unhandled Error: ${err.message}`, {
      path: req.path,
      method: req.method,
      stack: err.stack,
      errorObject: err
    });
  }

  // Handle specific error types
  if (error.name === 'JsonWebTokenError') {
    error = new AppError('Invalid token, please log in again', 401);
  }

  if (error.name === 'TokenExpiredError') {
    error = new AppError('Token has expired, please log in again', 401);
  }

  // Set default status code and message
  const statusCode = error instanceof AppError ? error.statusCode : 500;
  const message = error instanceof AppError ? error.message : 'Something went wrong!';

  res.status(statusCode).json({
    status: 'error',
    statusCode,
    message,
    // Include stack trace only in development environment for debugging
    stack: environment.nodeEnv === 'development' ? err.stack : undefined,
  });
};
```

File: project-{topic}-enhanced/src/middleware/ErrorHandlerMiddleware.cpp
```cpp
#include "ErrorHandlerMiddleware.hpp"
#include "../logger/Logger.hpp"
#include "../utils/AppException.hpp"

#include "crow.h"
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <string>

ErrorHandlerMiddleware::ErrorHandlerMiddleware() {
    // Constructor
}

// Called before the route handler. Not used for error handling directly.
void ErrorHandlerMiddleware::before_handle(crow::request& req, crow::response& res, Context& ctx) {
    // No specific logic needed before for error handling itself.
}

// Called after the route handler (or if an exception is thrown during before_handle/route_handle).
void ErrorHandlerMiddleware::after_handle(crow::request& req, crow::response& res, Context& ctx) {
    // If an exception was thrown and caught by Crow's internal mechanism,
    // the response will likely have a non-2xx status code.
    // This middleware aims to standardize the error response format.

    // If response status is not 2xx, or an exception was caught by Crow's default handler,
    // we want to reformat it.
    // Crow's exception handling flow is a bit tricky: if an exception is thrown in `before_handle`
    // or the route handler, Crow often catches it, sets a default error response, and then
    // calls `after_handle`. We intercept this to provide a consistent JSON format.

    // Check if the response is already an error (non-2xx)
    // or if ctx.exception_ptr has been set (indicating an uncaught exception)
    if (res.code >= 400 || ctx.exception_ptr) {
        // Retrieve exception information if available, or use generic message
        std::string errorMessage = "An unknown server error occurred.";
        int statusCode = 500;
        std::string errorDetails = "";

        if (ctx.exception_ptr) {
            try {
                std::rethrow_exception(ctx.exception_ptr);
            } catch (const AppException& e) {
                statusCode = e.getStatusCode();
                errorMessage = e.what();
                errorDetails = e.getDetails();
                Logger::error("ErrorHandler: Caught AppException ({}) for route {}: {}", statusCode, req.url, errorMessage);
            } catch (const std::exception& e) {
                // Catch any other standard exceptions
                errorMessage = e.what();
                statusCode = 500;
                Logger::error("ErrorHandler: Caught std::exception for route {}: {}", req.url, errorMessage);
            } catch (...) {
                // Catch all other unknown exceptions
                Logger::critical("ErrorHandler: Caught unknown exception for route {}.", req.url);
            }
            // Clear the exception pointer so Crow doesn't re-handle it
            ctx.exception_ptr = nullptr;
        } else {
            // If no exception_ptr but res.code is an error, use existing code and message (or generic)
            statusCode = res.code;
            if (res.body.empty() || res.body.find("error") == std::string::npos) {
                 // If Crow provided a non-JSON error or no body, provide a generic message
                 if (statusCode == 404) errorMessage = "Not Found.";
                 else if (statusCode == 401) errorMessage = "Unauthorized.";
                 else if (statusCode == 403) errorMessage = "Forbidden.";
                 else if (statusCode == 429) errorMessage = "Too Many Requests.";
                 else errorMessage = "An error occurred.";
            } else {
                // Try to parse Crow's default error body if it exists and is potentially JSON
                try {
                    nlohmann::json existingBody = nlohmann::json::parse(res.body);
                    if (existingBody.contains("message")) {
                        errorMessage = existingBody["message"].get<std::string>();
                    } else if (existingBody.contains("error")) {
                         errorMessage = existingBody["error"].get<std::string>();
                    }
                } catch (const nlohmann::json::parse_error&) {
                    // Not JSON, just use generic message
                }
            }
             Logger::error("ErrorHandler: Non-2xx response without explicit exception for route {}. Status: {}, Message: {}", 
                           req.url, statusCode, errorMessage);
        }

        // Construct a standardized JSON error response
        nlohmann::json errorResponse;
        errorResponse["status"] = "error";
        errorResponse["message"] = errorMessage;
        errorResponse["code"] = statusCode;
        if (!errorDetails.empty()) {
            errorResponse["details"] = errorDetails;
        }

        res.code = static_cast<crow::status>(statusCode);
        res.set_header("Content-Type", "application/json");
        res.body = errorResponse.dump();
    }
}
```